Source/Core/DolphinQt/Config/VerifyWidget.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt/Config/VerifyWidget.h"
+
+#include <memory>
+
+#include <QHeaderView>
+#include <QLabel>
+#include <QProgressDialog>
+#include <QPushButton>
+#include <QVBoxLayout>
+
+#include "DiscIO/Volume.h"
+#include "DiscIO/VolumeVerifier.h"
+
+VerifyWidget::VerifyWidget(std::shared_ptr<DiscIO::Volume> volume) : m_volume(std::move(volume))
+{
+  QVBoxLayout* layout = new QVBoxLayout(this);
+
+  m_problems = new QTableWidget(0, 2, this);
+  m_problems->setHorizontalHeaderLabels({tr("Problem"), tr("Severity")});
+  m_problems->horizontalHeader()->setSectionResizeMode(0, QHeaderView::Stretch);
+  m_problems->horizontalHeader()->setSectionResizeMode(1, QHeaderView::ResizeToContents);
+  m_problems->verticalHeader()->setSectionResizeMode(QHeaderView::ResizeToContents);
+  m_problems->verticalHeader()->hide();
+
+  m_summary_text = new QTextEdit(this);
+  m_summary_text->setReadOnly(true);
+
+  QPushButton* verify_button = new QPushButton(tr("Verify Integrity"));
+  connect(verify_button, &QPushButton::clicked, this, &VerifyWidget::Verify);
+
+  layout->addWidget(m_problems);
+  layout->addWidget(m_summary_text);
+  layout->addWidget(verify_button);
+
+  layout->setStretchFactor(m_problems, 5);
+  layout->setStretchFactor(m_summary_text, 2);
+
+  setLayout(layout);
+}
+
+void VerifyWidget::Verify()
+{
+  DiscIO::VolumeVerifier verifier(*m_volume);
+
+  QProgressDialog* progress =
+      new QProgressDialog(tr("Verifying"), tr("Cancel"), 0, verifier.GetMaxProgress(), this);
+  progress->setWindowTitle(tr("Verifying"));
+  progress->setWindowFlags(progress->windowFlags() & ~Qt::WindowContextHelpButtonHint);
+  progress->setMinimumDuration(500);
+  progress->setWindowModality(Qt::WindowModal);
+
+  verifier.Start();
+  while (verifier.GetProgress() != verifier.GetMaxProgress())
+  {
+    progress->setValue(verifier.GetProgress());
+    if (progress->wasCanceled())
+      return;
+
+    verifier.Process();
+  }
+  verifier.Finish();
+
+  DiscIO::VolumeVerifier::Result result = verifier.GetResult();
+  progress->setValue(verifier.GetProgress());
+
+  m_summary_text->setText(QString::fromStdString(result.summary_text));
+
+  m_problems->setRowCount(static_cast<int>(result.problems.size()));
+  for (int i = 0; i < m_problems->rowCount(); ++i)
+  {
+    const DiscIO::VolumeVerifier::Problem problem = result.problems[i];
+
+    QString severity;
+    switch (problem.severity)
```

Comment:
Do we want to use shading or text color to indicate severity as well, or should we just leave it as normal text?

Reply:
@MayImilae, any thoughts?